cvmfs/magic_xattr.cc

```diff
@@ -482,42 +514,93 @@ std::string ProxyMagicXattr::GetValue() {
   }
 }
 
-static std::string ListProxy(download::DownloadManager *dm) {
+static void ListProxy(download::DownloadManager *dm,
+                      std::vector<std::string> *result_pages) {
   vector< vector<download::DownloadManager::ProxyInfo> > proxy_chain;
   unsigned current_group;
   dm->GetProxyInfo(&proxy_chain, &current_group, NULL);
   std::string buf = "";
+  std::string fulbuf = "";
   for (unsigned int i = 0; i < proxy_chain.size(); i++) {
     for (unsigned int j = 0; j < proxy_chain[i].size(); j++) {
       buf += proxy_chain[i][j].url;
       buf += "\n";
     }
+
+    if (buf.size() > BaseMagicXattr::kMaxCharsPerPage) {
+      result_pages->push_back(buf);
+      buf = "";
+    }
+  }
+
+  if (buf.size() > 0) {
+    result_pages->push_back(buf);
   }
-  return buf;
 }
 
-std::string ProxyListMagicXattr::GetValue() {
-  return ListProxy(xattr_mgr_->mount_point()->download_mgr());
+std::string ProxyListMagicXattr::GetValue(uint32_t requested_page) {
+  result_pages_.clear();
+  ListProxy(xattr_mgr_->mount_point()->download_mgr(), &result_pages_);
+
+  return BaseMagicXattr::GetValue(requested_page);
 }
 
-std::string ProxyListExternalMagicXattr::GetValue() {
-  return ListProxy(xattr_mgr_->mount_point()->external_download_mgr());
+std::string ProxyListExternalMagicXattr::GetValue(uint32_t requested_page) {
+  result_pages_.clear();
+  ListProxy(xattr_mgr_->mount_point()->external_download_mgr(), &result_pages_);
+
+  return BaseMagicXattr::GetValue(requested_page);
 }
 
 bool PubkeysMagicXattr::PrepareValueFenced() {
   pubkeys_ = xattr_mgr_->mount_point()->signature_mgr()->GetActivePubkeys();
   return true;
 }
 
-std::string PubkeysMagicXattr::GetValue() {
-  return pubkeys_;
+std::string PubkeysMagicXattr::GetValue(uint32_t requested_page) {
+  result_pages_.clear();
+
+  if (pubkeys_.size() <= kMaxCharsPerPage) {
+    result_pages_.push_back(pubkeys_);
+  } else {
+    // we need to create our own pages, no need to do it in the fuse fence
+    result_pages_.push_back("pubkey size " + StringifyUint(pubkeys_.size()));
+    uint64_t start_idx = 0;
+    uint64_t end_idx = kMaxCharsPerPage - 400;  // find a hit for the end of a
```

Comment:
What if we don't find the marker within these bounds?

I think it is fine to cut the attribute in-between a key. That makes our code simpler.

Reply:
the else case is only hit if we have at least a string of all pubkeys >40000 symbols. otherwise we just return the string we get (can be empty)
i found it much nicer to cut after `-----END PUBLIC KEY-----` as you might want to compare pub keys directly (with an easy `ctrl+c`) and it is way nicer to copy. i dont think this extra logic creates so much overhead. it is already written and i tested it by hand.

Reply:
The extended attribute interface is open to regular users, and unforeseen requests have crashed the client in the past. Parsing the key files correctly will require more (automated) testing. I doubt that it is worthwhile. Scripts using the attribute anyway need to combine information from all pages. If we cut in the middle of the key, users need to combine the pages without extra whitespace (newline etc.). In the documentation of the attribute, we can suggest a command to do so.